amundsen_application/static/js/components/common/ResourceList/PaginatedResourceList/index.spec.tsx

```diff
@@ -0,0 +1,119 @@
+import * as React from 'react';
+import Pagination from 'react-js-pagination';
+
+import { shallow } from 'enzyme';
+
+import { ResourceType } from 'interfaces';
+import ResourceListItem from 'components/common/ResourceListItem/index';
+
+import PaginatedResourceList, { PaginatedResourceListProps } from './';
+import * as CONSTANTS from '../constants';
+
+describe('PaginatedResourceList', () => {
+  const setStateSpy = jest.spyOn(PaginatedResourceList.prototype, 'setState');
+  const setup = (propOverrides?: Partial<PaginatedResourceListProps>) => {
+    const props: PaginatedResourceListProps = {
+      allItems: [
+        { type: ResourceType.table },
+        { type: ResourceType.table },
+        { type: ResourceType.table },
+        { type: ResourceType.table },
+        { type: ResourceType.table },
+        { type: ResourceType.table },
+      ],
+      itemsPerPage: 4,
+      source: 'testSource',
+      ...propOverrides
+    };
+    const wrapper = shallow<PaginatedResourceList>(<PaginatedResourceList {...props} />);
+    return { props, wrapper };
+  };
+
+  describe('componentDidUpdate', () => {
+    it('updates activePage state if the activePage is out of bounds by multiple pages', () => {
+      const wrapper = setup({ itemsPerPage: 2 }).wrapper;
+      wrapper.setState({ activePage: 2 });
+      const componentDidUpdateSpy = jest.spyOn(wrapper.instance(), 'componentDidUpdate');
+      wrapper.setProps({
+        allItems: [
+          { type: ResourceType.table },
+          { type: ResourceType.table },
+        ]
+      });
+      expect(componentDidUpdateSpy).toHaveBeenCalled();
+      expect(wrapper.state().activePage).toEqual(0);
+    })
+
+    it('updates activePage state if the activePage is out of bounds by one page', () => {
+      const wrapper = setup({ itemsPerPage: 2 }).wrapper;
+      wrapper.setState({ activePage: 2 });
+      const componentDidUpdateSpy = jest.spyOn(wrapper.instance(), 'componentDidUpdate');
+      wrapper.setProps({
+        allItems: [
+          { type: ResourceType.table },
+          { type: ResourceType.table },
+          { type: ResourceType.table },
+        ]
+      })
+      expect(componentDidUpdateSpy).toHaveBeenCalled();
+      expect(wrapper.state().activePage).toEqual(1);
+    })
+
+    it('does not update activePage if new values are not out of bounds', () => {
+      const wrapper = setup({ itemsPerPage: 2 }).wrapper;
+      wrapper.setState({ activePage: 2 });
+      wrapper.setProps({
+        allItems: [
+          { type: ResourceType.table },
+          { type: ResourceType.table },
+          { type: ResourceType.table },
+          { type: ResourceType.table },
+          { type: ResourceType.table },
+        ]
+      })
+      expect(wrapper.state().activePage).toEqual(2);
+    })
+  });
```

Comment:
Given that the result of this logic is different values to the activePage state value, could we check this on the rendered results instead?

Reply:
Could you elaborate I'm not sure I follow?

Reply:
Yes, let me expand on it here.

In a similar way as in my comment below, I have experienced the issues that come from checking for implementation details on tests. 

In this case, we are rendering a component with a giving state, simulating a change of props and making sure that the new state is what we expect. This makes sense, and it is not a wrong test in any case. Actually, following the "London School of testing", this is exactly what they propose. They prefer deep tests that are strict but not flexible.

However, the way I have been using and recommend to use is a bit less detailed, but more flexible. This way would be to only check for the "public effects" of the actions. In this case, that would be what gets rendered in the DOM. That would mean that we would set up the component the same way, we would act solely on the 'onPagination' method and will check the results on the DOM. (In this case, as commented below, it would mean we depend on the Pagination component to render what we need). The benefits of this way ("The Chicago School" of testing) is that our tests would be more flexible and break less, at the cost of less strict checking to the implementation details. 

A good example we can think of now is the refactor to hooks. The test we are adding now would break with hooks, but if we check only the rendered result, they would still pass.

That said, I don't actually recommend to remove this tests. When I see this tests on a code base, I prefer them than no tests, and most of the times it is not worth to rewrite them. So what I suggest is to leave them here, but the moment that they break, we should remove them (the same for the next test).

Sorry for the wall of text!!

Reply:
This makes sense given the example of hooks, which as I think I mentioned I' m unfamiliar but have a goal to learn more about. Let's continue to talk about this offline. None of our methods are set in stone, we essentially learn as we go discussing different perspectives on the best way to approach things. And then if we do shift practices, for unit tests at least we just make those changes moving forward vs. trying to retroactively fix update hundreds of tests. 

So in short, let's keep discussing and feel free to write tests from your perspective.

Reply:
Totally, makes sense! 
Thanks!